### PR TITLE
Init

### PR DESCRIPTION
**Pull request by jstucke**
Initial Commit

### CODE REVIEW

File: code/codescanner.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+from plugins.mime_blacklists import MIME_BLACKLIST_COMPRESSED, MIME_BLACKLIST_NON_EXECUTABLE
+
+
+class AnalysisPlugin(AnalysisBasePlugin):
+
+    NAME = 'codescanner'
+    DESCRIPTION = 'scan unknown binaries for executable code'
+    VERSION = '1.0.0'
+    DEPENDENCIES = ['file_type']
+    MIME_BLACKLIST = MIME_BLACKLIST_NON_EXECUTABLE + MIME_BLACKLIST_COMPRESSED
```
Comment (rhelmke) [inline code/codescanner.py:14]: `MIME_BLACKLIST_NON_EXECUTABLE`.

I'm wondering if running the tool on already identified executable files might still be reasonable. The tool provides a lot of interesting information and the plots might be a valuable contribution for further in-depth analysis. Think about a packed ELF files, you could probably see that pretty fast using codescanner plots.

What do you think? Lets talk! :-)

File: test/test_codescanner.py [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+class TestAnalysisPluginCryptoHints(AnalysisPluginTest):
+
+    PLUGIN_NAME = 'codescanner'
+    PLUGIN_CLASS = AnalysisPlugin
+
+    def test_process_object(self):
+        for test_file, expected_result in [
+            ('amd64.elf', 'Intel-64'),
+            ('arm.elf', 'Arm-32-le'),
+        ]:
```
Comment (rhelmke) [inline test/test_codescanner.py:21]: Do we have any test file that contains hidden executable data? That could be a nice addition to the testing here. :-)

File: install.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    def build(self):
+        build_dir = self.base_path / 'build' / 'Codescanner'
+        run_cmd_with_logging(
+            f'git clone https://github.com/fkie-cad/Codescanner.git -o Codescanner', raise_error=True
+        )
+        with OperateInDirectory(build_dir, remove=True):
+            command = 'python3 setup.py install'
+            if not is_virtualenv():
+                logging.warning("no venv")
+                command = f'sudo -EH {command}'
```
Comment (rhelmke) [inline install.py:30]: This does not work for non-virtual installations.

The install succeeds, but the package is only accessible by the root user. Running fact as non-root yields:

```python3
**  File "/home/fact/FACT_core/src/./plugins/analysis/codescanner_plugin/routes/routes.py", line 3, in <module>
    from codescanner_analysis import CodescannerAnalysisData
ModuleNotFoundError: No module named 'codescanner_analysis**
```

And, indeed, if I try to import `codescanner_analysis` using the same user, it doesn't work:

```bash
$ python3 -c "import codescanner_analysis"
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ModuleNotFoundError: No module named 'codescanner_analysis'
$
```

However, using root everything is fine:

```bash
$ sudo python3 -c "import codescanner_analysis"
$
```

**Review by rhelmke — CHANGES_REQUESTED** [review]

File: code/codescanner.py (hunk 1 above)
Comment (rhelmke) [reply code/codescanner.py:14]: We talked about it. The blacklist stays as-is.

File: code/codescanner.py [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+                key: {
+                    'size': binary.sizes.get(key),
+                    'regions': binary.regions.get(key),
+                }
+            })
+
+        if result['architecture']['Full'] is not None:
+            result['summary'] = [result['architecture']['Full']]
+        file_object.processed_analysis[self.NAME] = result
+        return file_object
```
[the remark this answers is not in the document]
Comment (jstucke) [reply code/codescanner.py:37]: For very small files and files without code regions everything should simply be `None`. I was unable to generate errors in this part.

File: routes/routes.py [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+        binary = CodescannerAnalysisData(self.fso.generate_path_from_uid(uid))
+        image = binary.plot_to_buffer(dpi=100, plot_type=binary.BYTE_PLOT)
+        encoded_img = b64encode(image).decode()
+        template_str = f'<img style="max-width:100%;" src="data:image/png;base64,{encoded_img}" width="1024px" />'
+        return render_template_string(template_str)
+
+    @roles_accepted(*PRIVILEGES['view_analysis'])
+    def _get_color_map(self, uid):
+        binary = CodescannerAnalysisData(self.fso.generate_path_from_uid(uid))
+        image = binary.plot_to_buffer(dpi=100, plot_type=binary.COLOR_MAP)
```
[the remark this answers is not in the document]
Comment (jstucke) [reply routes/routes.py:33]: Good catch. I added error handling for this case

File: test/test_codescanner.py (hunk 2 above)
Comment (jstucke) [reply test/test_codescanner.py:21]: I added a file containing a raw bare metal binary between some random data. Is that what you had in mind?

File: code/codescanner.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (rhelmke) [reply code/codescanner.py:49]: Okay, I see. Thanks!

File: requirements.txt [hunk 6]
```diff
@@ -0,0 +1 @@
+git+https://github.com/fkie-cad/Codescanner.git
```
Comment (rhelmke) [inline requirements.txt:1]: Should we pin the commit? 8)

```suggestion
git+https://github.com/fkie-cad/Codescanner.git@02262b6345dfb156ddcc822f4e84586a7c6c1a7a
```

File: install.py (hunk 3 above)
Comment (rhelmke) [reply install.py]: Great, works flawlessly now

**Pull request merged**